### PR TITLE
feat: direct BigQueryAnalyticsHubDataExchange

### PR DESCRIPTION
**Pull request by acpana**
Converts *v1alpha* `BigQueryAnalyticsHubDataExchange` as a direct resource. Note we are choosing to treat this as a special case and first do the conversion and then add the mockgcp as the TF controller is using a proto version that is not published for this service.

### CODE REVIEW

File: apis/bigqueryanalyticshub/v1alpha1/bigqueryanalyticshubdataexchange_types.go [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+import (
+	refs "github.com/GoogleCloudPlatform/k8s-config-connector/apis/refs/v1beta1"
+
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/apis/k8s/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var BigQueryAnalyticsHubDataExchangeGVK = GroupVersion.WithKind("BigQueryAnalyticsHubDataExchange")
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
```
Comment (yuwenma) [inline apis/bigqueryanalyticshub/v1alpha1/bigqueryanalyticshubdataexchange_types.go:26]: nit: cleanup

**Review by yuwenma — COMMENTED** [review]
/hold 

We'd want to have the mock gcp coverage first for Alpha resource. https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/master/docs/develop-resources/scenarios/migrate-tf-resource-alpha.md

File: apis/bigqueryanalyticshub/v1alpha1/bigqueryanalyticshubdataexchange_types.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+	// Optional. Documentation describing the data exchange.
+	Documentation *string `json:"documentation,omitempty"`
+
+	// Optional. Base64 encoded image representing the data exchange. Max
+	//  Size: 3.0MiB Expected image dimensions are 512x512 pixels, however the API
+	//  only performs validation on size of the encoded data. Note: For byte
+	//  fields, the content of the fields are base64-encoded (which increases the
+	//  size of the data by 33-36%) when using JSON on the wire.
+	Icon *string `json:"icon,omitempty"`
```
Comment (justinsb) [inline apis/bigqueryanalyticshub/v1alpha1/bigqueryanalyticshubdataexchange_types.go:61]: I'm wondering if we should NOTYET this.  The idea of 3MB of encoded image data scares me :-)

File: config/crds/resources/apiextensions.k8s.io_v1_customresourcedefinition_bigqueryanalyticshubdataexchanges.bigqueryanalyticshub.cnrm.cloud.google.com.yaml [hunk 3]
```diff
@@ -8,18 +8,15 @@ metadata:
     cnrm.cloud.google.com/managed-by-kcc: "true"
     cnrm.cloud.google.com/stability-level: alpha
     cnrm.cloud.google.com/system: "true"
-    cnrm.cloud.google.com/tf2crd: "true"
```
Comment (justinsb) [inline config/crds/resources/apiextensions.k8s.io_v1_customresourcedefinition_bigqueryanalyticshubdataexchanges.bigqueryanalyticshub.cnrm.cloud.google.com.yaml:11]: Ah, I guess we can't NOTYET it because it was defined in Terraform .... it is alpha, but if we choose to remove it, let's do that in a different PR

File: config/crds/resources/apiextensions.k8s.io_v1_customresourcedefinition_bigqueryanalyticshubdataexchanges.bigqueryanalyticshub.cnrm.cloud.google.com.yaml [hunk 4]
```diff
@@ -8,18 +8,15 @@ metadata:
...
   name: bigqueryanalyticshubdataexchanges.bigqueryanalyticshub.cnrm.cloud.google.com
 spec:
   group: bigqueryanalyticshub.cnrm.cloud.google.com
   names:
     categories:
     - gcp
     kind: BigQueryAnalyticsHubDataExchange
+    listKind: BigQueryAnalyticsHubDataExchangeList
     plural: bigqueryanalyticshubdataexchanges
-    shortNames:
```
Comment (justinsb) [inline config/crds/resources/apiextensions.k8s.io_v1_customresourcedefinition_bigqueryanalyticshubdataexchanges.bigqueryanalyticshub.cnrm.cloud.google.com.yaml:20]: Nit: Can you add these - you can add annotations to the type definition.

Not blocking, and maybe we should automate generating them

File: apis/bigqueryanalyticshub/v1alpha1/bigqueryanalyticshubdataexchange_types.go [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// BigQueryAnalyticsHubDataExchangeSpec defines the desired state of BigQueryAnalyticsHubDataExchange
+// +kcc:proto=google.cloud.bigquery.analyticshub.v1.DataExchange
+type BigQueryAnalyticsHubDataExchangeSpec struct {
+	// Output only. The resource name of the data exchange.
+	//  e.g. `projects/myproject/locations/US/dataExchanges/123`.
+	Name *string `json:"name,omitempty"`
```
Comment (justinsb) [inline apis/bigqueryanalyticshub/v1alpha1/bigqueryanalyticshubdataexchange_types.go:34]: Nit: if this is output only, I don't think we want it in spec.  I guess it's in spec because of TF?  If so, let's remove it in a separate PR (we're allowed to do that because alpha)

File: pkg/clients/generated/apis/bigqueryanalyticshub/v1alpha1/bigqueryanalyticshubdataexchange_types.go [hunk 6]
```diff
@@ -36,59 +36,73 @@ import (
...
 	ObservedGeneration *int64 `json:"observedGeneration,omitempty"`
+
+	/* ObservedState is the state of the resource as most recently observed in GCP. */
+	// +optional
+	ObservedState *DataexchangeObservedStateStatus `json:"observedState,omitempty"`
 }
 
 // +genclient
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
-// +kubebuilder:resource:categories=gcp,shortName=gcpbigqueryanalyticshubdataexchange;gcpbigqueryanalyticshubdataexchanges
```
Comment (justinsb) [inline pkg/clients/generated/apis/bigqueryanalyticshub/v1alpha1/bigqueryanalyticshubdataexchange_types.go:89]: Ah - here are the annotations we need :-)

File: pkg/test/resourcefixture/testdata/basic/bigqueryanalyticshub/v1alpha1/bigqueryanalyticshubdataexchange/bigqueryanalyticshubdataexchangebasic/create.yaml [hunk 7]
```diff
@@ -15,7 +15,7 @@
 apiVersion: bigqueryanalyticshub.cnrm.cloud.google.com/v1alpha1
 kind: BigQueryAnalyticsHubDataExchange
 metadata:
-  name: bigqueryanalyticshubdataexchange-${uniqueId}
+  name: bigqueryanalyticshubdataexchange${uniqueId}
```
Comment (justinsb) [inline pkg/test/resourcefixture/testdata/basic/bigqueryanalyticshub/v1alpha1/bigqueryanalyticshubdataexchange/bigqueryanalyticshubdataexchangebasic/create.yaml:18]: Out of interest, why the change?

**Comment by justinsb** [comment]
(Holding in case you want to address anything here)

File: pkg/controller/direct/bigqueryanalyticshub/dataexchange_externalresource.go [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+// FullyQualifiedName returns both parent and resource ID in the full url format.
+func (c *DataExchangeIdentity) FullyQualifiedName() string {
+	return fmt.Sprintf("%s/dataExchanges/%s", c.Parent.String(), c.DataExchange)
+}
+
+// AsExternalRef builds a externalRef from a DataExchange
+func (c *DataExchangeIdentity) AsExternalRef() *string {
+	e := serviceDomain + "/" + c.FullyQualifiedName()
```
[the remark this answers is not in the document]
Comment (acpana) [reply pkg/controller/direct/bigqueryanalyticshub/dataexchange_externalresource.go:44]: I think this mostly came from the contorllerbuilder tool

File: pkg/test/resourcefixture/testdata/basic/bigqueryanalyticshub/v1alpha1/bigqueryanalyticshubdataexchange/bigqueryanalyticshubdataexchangebasic/create.yaml (hunk 7 above)
Comment (acpana) [reply pkg/test/resourcefixture/testdata/basic/bigqueryanalyticshub/v1alpha1/bigqueryanalyticshubdataexchange/bigqueryanalyticshubdataexchangebasic/create.yaml:18]: I've started using the name as the dataExchangeId and it only allows a-z, 0-9 characters IIRC

File: apis/bigqueryanalyticshub/v1alpha1/bigqueryanalyticshubdataexchange_types.go (hunk 5 above)
Comment (acpana) [reply apis/bigqueryanalyticshub/v1alpha1/bigqueryanalyticshubdataexchange_types.go:31]: Going to make this change as a separate PR

**Comment by justinsb** [comment]
Thanks @acpana 

/approve
/lgtm
